packages/std/src/math/conversion.rs

```diff
@@ -0,0 +1,104 @@
+/// Grows a big endian signed integer to a bigger size.
+/// See <https://en.wikipedia.org/wiki/Sign_extension>
+pub const fn grow_be_int<const INPUT_SIZE: usize, const OUTPUT_SIZE: usize>(
+    input: [u8; INPUT_SIZE],
+) -> [u8; OUTPUT_SIZE] {
+    debug_assert!(INPUT_SIZE <= OUTPUT_SIZE);
+    // check if sign bit is set
+    let mut output = if input[0] & 0b10000000 != 0 {
+        // negative number is filled up with 1s
+        [0b11111111u8; OUTPUT_SIZE]
+    } else {
+        [0u8; OUTPUT_SIZE]
+    };
+    let mut i = 0;
+
+    // copy input to the end of output
+    // copy_from_slice is not const, so we have to do this manually
+    while i < INPUT_SIZE {
+        output[OUTPUT_SIZE - INPUT_SIZE + i] = input[i];
+        i += 1;
+    }
+    output
+}
+
+/// Shrinks a big endian signed integer to a smaller size.
+/// This is the opposite operation of sign extension.
+pub fn shrink_be_int<const INPUT_SIZE: usize, const OUTPUT_SIZE: usize>(
+    input: [u8; INPUT_SIZE],
+) -> Option<[u8; OUTPUT_SIZE]> {
+    debug_assert!(INPUT_SIZE >= OUTPUT_SIZE);
+
+    // check bounds
+    if input[0] & 0b10000000 != 0 {
+        // a negative number should start with only 1s, otherwise it's too small
+        for i in &input[0..(INPUT_SIZE - OUTPUT_SIZE)] {
+            if *i != 0b11111111u8 {
+                return None;
+            }
+        }
+    } else {
+        // a positive number should start with only 0s, otherwise it's too large
+        for i in &input[0..(INPUT_SIZE - OUTPUT_SIZE)] {
+            if *i != 0u8 {
+                return None;
+            }
+        }
+    }
+
+    // Now, we can just copy the last bytes
+    let mut output = [0u8; OUTPUT_SIZE];
+    output.copy_from_slice(&input[(INPUT_SIZE - OUTPUT_SIZE)..]);
+    Some(output)
+}
+
```

Comment:
Would be nice to try to upstream those. I.e. go from `BInt::<X>` to `BInt::<Y>`. Then you can also operate in 8 steps which should be faster.

Reply:
Actually, after looking into this, there seems to already be a way of casting them using `bnum`'s `CastFrom` or `As` trait, but it needs nightly features to be `const`.

Reply:
Alright cool. This solution is good anyways. I just want to be nice and upstream ideas and code where possible.